File: Advent of Code/Challenge Solutions/Year 2021/ChallengeSolution06.cs
// Task: https://adventofcode.com/2021/day/6

using Advent_of_Code.Utilities;

namespace Advent_of_Code.Challenge_Solutions.Year_2021;

public class ChallengeSolution06(IConsole console) : ChallengeSolution(console)
{
    public override void SolveFirstPart()
    {
        List<int> lanternfish = new List<int>(Array.ConvertAll(Reader.ReadLines(this)[0].Split(",", StringSplitOptions.RemoveEmptyEntries), int.Parse));

        int days = 0;

        while (days < 80)
        {
            List<int> newFish = new List<int>();
            for (int i = 0; i < lanternfish.Count; i++)
            {
                if (lanternfish[i] == 0)
                {
                    lanternfish[i] = 6;
                    newFish.Add(8);
                }
                else
                    lanternfish[i]--;
            }

            foreach (var fish in newFish)
                lanternfish.Add(fish);

            days++;
        }

        Console.WriteLine(lanternfish.Count);
    }

    public override void SolveSecondPart()
    {
        long[] fish = new long[9];
        List<int> initialFish = new List<int>(Array.ConvertAll(Reader.ReadLines(this)[0].Split(",", StringSplitOptions.RemoveEmptyEntries), int.Parse));
        foreach (var f in initialFish)
            fish[f]++;

        int days = 0;
        while (days < 256)
        {
            long birthingFish = fish[0];
            for (int i = 0; i < 8; i++)
                fish[i] = fish[i + 1];
            fish[8] = 0;

            fish[6] += birthingFish;
            fish[8] += birthingFish;

            days++;
        }

        long sum = 0;
        foreach (var f in fish)
            sum += f;

        Console.WriteLine(sum);
    }
}

File: Advent of Code.FSharp/Challenge Solutions/Year 2025/ChallengeSolution06.fs
// Task: https://adventofcode.com/2025/day/6

namespace Advent_of_Code.FSharp.Challenge_Solutions.Year_2025

open System
open Advent_of_Code.Shared
open Advent_of_Code.Shared.Utilities

module ChallengeSolution06 =

    let whitespace = ' '

    type Operation =
        | Addition
        | Multiplication

    module Operation =

        let ofString operator =
            match operator with
            | "+" -> Addition
            | "*" -> Multiplication
            | _ -> failwith $"Unknown operator {operator}"

        let ofChar operator =
            ofString $"{operator}"

        let inline apply a op b =
            match op with
            | Addition       -> a + b
            | Multiplication -> a * b

        let getIdentity op =
            match op with
            | Addition       -> 0
            | Multiplication -> 1

    type Problem = {
        Numbers: uint64 list
        Operation: Operation
    }

    module Problem =
        let compute problem =
            let op = problem.Operation
            let identity = (Operation.getIdentity op |> uint64)
            problem.Numbers
            |> List.fold (fun acc x -> Operation.apply acc op x) identity

    let removeLast list =
        match list with
        | [] -> []
        | _ -> list |> List.rev |> List.tail |> List.rev

    type Solution(getInput: unit -> string list, writeLine: string -> unit) =

        let readProblemsAsHuman() =
            let rec convertColumnToProblem numbers columnElements =
                match columnElements with
                | [] -> failwith "Line has no elements!"
                | [op] -> {
                        Numbers = numbers
                        Operation = op |> Operation.ofString
                    }
                | numberString :: remainingElements ->
                    let numbers = (numberString |> UInt64.Parse) :: numbers
                    remainingElements |> convertColumnToProblem numbers

            getInput()
            |> List.map (fun line -> 
                line.Split(whitespace, StringSplitOptions.RemoveEmptyEntries)
                |> List.ofArray)
            |> List.transpose
            |> List.map (convertColumnToProblem [])

        let readProblemsAsCephalopod() =
            let groupLinesByProblem lines =
                let groupsReversed, current =
                    lines
                    |> List.fold (fun (problemStrings, current) line ->
                        if line |> List.forall (fun (c: char) -> c = whitespace) then
                            match current with
                            | [] -> (problemStrings, [])
                            | _  -> (List.rev current :: problemStrings, [])
                        else 
                            (problemStrings, line :: current)
                    ) ([], [])

                let groupsReversed =
                    match current with
                    | [] -> groupsReversed
                    | _ -> List.rev current :: groupsReversed

                List.rev groupsReversed

            let lineGroupToProblem (group: char list list) =
                let operation = group[0] |> List.last |> Operation.ofChar
                let numbers =
                    group
                    |> List.map (fun line -> 
                        line
                        |> removeLast
                        |> List.map (
                            function
                            | c when c >= '0' && c <= '9' -> (int c - int '0') |> Some
                            | _ -> None))
                    |> List.map (
                        List.fold (fun acc digit -> 
                            match digit with
                            | Some d -> acc * 10 + d
                            | None   -> acc) 0)
                    |> List.map uint64

                { 
                    Numbers = numbers
                    Operation = operation 
                }

            getInput()
            |> List.map Seq.toList
            |> List.transpose
            |> groupLinesByProblem
            |> List.map lineGroupToProblem

        let computeGrandTotal (readProblems: unit -> Problem list) =
            readProblems() 
            |> List.sumBy Problem.compute
        
        member _.SolveFirstPart() =
            let total = computeGrandTotal readProblemsAsHuman
            writeLine $"Grand total: {total}"

        member _.SolveSecondPart() : unit =
            let total = computeGrandTotal readProblemsAsCephalopod
            writeLine $"Grand total: {total}"

    type ChallengeSolution06(console: IConsole, reader: ISolutionReader<ChallengeSolution06>) =
        inherit ChallengeSolution<ChallengeSolution06>(console, reader)

        let solution = Solution(reader.ReadLines >> List.ofArray, console.WriteLine)

        override this.SolveFirstPart() = solution.SolveFirstPart()
        override this.SolveSecondPart() = solution.SolveSecondPart()
